woocommerce/woocommerce
# [HPOS CLI] Add support for backfilling specific properties or metadata

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:




In #44281 we introduced the `wp wc hpos backfill` command which allows backfilling an order from/to the HPOS or posts datastore, regardless of which datastore is currently active. This allows merchants to fix one off errors where some data might've been modified outside of the CRUD layer.

While this works ok, sometimes there might be a need to sync just some specific metadata or properties from/to either datastore instead of the whole order. This PR introduces `--meta-keys` and `--props` arguments to `wp wc hpos backfill` which do precisely that.
While ideally orders would be backfilled in their entirety, this at least provides a last resort for some possible tricky cases.

Closes #41910.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure HPOS is set as datastore in WC > Settings > Advanced > Features and also that sync (compatibility mode) is disabled.
   **Note:** It shouldn't really matter which datastore is set and we'll test syncing both ways, but this will make testing instructions more straightforward.
2. Make sure your site has at least one order with billing details. Take note of its ID.
3. Run `wp wc cot sync` to ensure that things are 100% synced between datastores.
4. Open the order from step 2 for editing:
   - Change its status to any status.
   - Change the billing first name to any value you like.
   - Add some metadata using the metadata metabox ("Custom Fields") and make a note of the "Name" you used for the field.
5. Run `wp wc hpos diff <order_id>` and confirm that the changes made above are listed as differences.
6. Run `wp wc hpos backfill <order_id> --from=hpos --to=posts --props=billing_first_name` to backfill the billing first name alone.
7. Run `wp wc hpos diff <order_id>` again and confirm that the status is no longer listed as being different. This proves that the `--props` arg limited the backfill to only that prop (leaving "status" and meta alone).
8. Let's add some metadata to the posts version with `wp post meta update <order_id> my_posts_meta my_meta_value`.
9. Run `wp wc hpos diff <order_id>` to confirm that the new meta is listed as a difference (with a value only on the 'post' side).
10. Let's migrate `my_posts_meta` from the post to the HPOS order by running `wp wc hpos backfill <order_id> --from=posts --to=hpos --meta_keys=my_posts_meta`. Run `wp wc hpos diff <order_id>` to confirm that this difference has now disappeared.
11. Now let's migrate the HPOS meta you added on step 4 by running `wp wc hpos backfill <order_id> --from=posts --to=hpos --meta_keys=<meta_key>`. Again, verify that the difference is gone with `wp wc hpos diff <order_id>`.
12. Finally, let's backfill a property (in this case, the status, which we've left alone thus far) from posts to HPOS: `wp wc hpos backfill <order_id> --from=posts --to=hpos --props=status`. Verify once more with `wp wc hpos diff <order_id>`.
13. Thank you for following such long testing instructions. You're a beautiful person 😸.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
